File: src/engines/baseline/greedy.py
"""The coarse card preferences and family selection shared by both greedy bots."""

from random import Random

from bluff_mau_mau import Draw, Move, PlayCard, Skip, _contribution
from ..observation import Observation


def best_play(observation: Observation, plays: list[PlayCard], rng: Random) -> PlayCard:
    # ponytail: fixed priorities; replace with evaluation/search when comparisons justify it.
    def score(move):
        declared, actual = move.declared_card, move.actual_card
        attack = bool(_contribution(declared) or declared.rank == "A")
        ordinary_actual = not (_contribution(actual) or actual.rank in ("A", "Q"))
        suit_count = sum(card.suit == move.chosen_suit for card in observation.hand
                         if card != actual)
        return (
            actual != declared and declared in observation.hand,
            attack if observation.opponent_count <= 2 else not (attack or declared.rank == "Q"),
            ordinary_actual,
            suit_count,
        )

    scores = [score(move) for move in plays]
    best = max(scores)
    return rng.choice([move for move, value in zip(plays, scores) if value == best])


def greedy_turn(observation: Observation, moves: tuple[Move, ...], rng: Random,
                bluff: int = 0, no_truth_bluff: int = 0) -> Move:
    truthful = [move for move in moves if isinstance(move, PlayCard)
                and move.actual_card == move.declared_card]
    bluffs = [move for move in moves if isinstance(move, PlayCard)
              and move.actual_card != move.declared_card]
    passes = [move for move in moves if isinstance(move, (Draw, Skip))]
    if truthful:
        plays = bluffs if bluffs and bluff and rng.random() < bluff / 100 else truthful
    elif bluffs and (not passes or no_truth_bluff and rng.random() < no_truth_bluff / 100):
        plays = bluffs
    else:
        return passes[0]
    return best_play(observation, plays, rng)

File: src-rust/src/engines/baseline/greedy.rs
use crate::engines::observation::Observation;
use crate::game::{Move, contribution};
use crate::rng::PythonRandom;

fn best_play(observation: &Observation, plays: &[Move], rng: &mut PythonRandom) -> Move {
    // ponytail: fixed priorities; evaluation/search can replace these after baseline comparisons.
    let score = |action: &Move| {
        let Move::Play {
            actual_card,
            declared_card,
            chosen_suit,
        } = action
        else {
            unreachable!()
        };
        let attack = contribution(*declared_card) > 0 || declared_card.rank() == 7;
        let ordinary = contribution(*actual_card) == 0 && ![7, 5].contains(&actual_card.rank());
        let suit_count = observation
            .hand
            .iter()
            .filter(|card| **card != *actual_card && Some(card.suit()) == *chosen_suit)
            .count();
        (
            *actual_card != *declared_card && observation.hand.contains(declared_card),
            if observation.opponent_count <= 2 {
                attack
            } else {
                !(attack || declared_card.rank() == 5)
            },
            ordinary,
            suit_count,
        )
    };
    let best = plays.iter().map(score).max().unwrap();
    let tied: Vec<Move> = plays
        .iter()
        .copied()
        .filter(|action| score(action) == best)
        .collect();
    tied[rng.randbelow(tied.len())]
}

pub fn greedy_turn(
    observation: &Observation,
    moves: &[Move],
    rng: &mut PythonRandom,
    bluff: u8,
    no_truth_bluff: u8,
) -> Result<Move, String> {
    let mut truthful = Vec::new();
    let mut bluffs = Vec::new();
    let mut pass = None;
    for &action in moves {
        match action {
            Move::Play {
                actual_card,
                declared_card,
                ..
            } => {
                if actual_card == declared_card {
                    truthful.push(action);
                } else {
                    bluffs.push(action);
                }
            }
            Move::Draw | Move::Skip if pass.is_none() => pass = Some(action),
            _ => {}
        }
    }
    let plays = if !truthful.is_empty() {
        if !bluffs.is_empty() && bluff != 0 && rng.random() < f64::from(bluff) / 100.0 {
            &bluffs
        } else {
            &truthful
        }
    } else if !bluffs.is_empty()
        && (pass.is_none()
            || no_truth_bluff != 0 && rng.random() < f64::from(no_truth_bluff) / 100.0)
    {
        &bluffs
    } else {
        return pass.ok_or_else(|| "No surviving play or pass".into());
    };
    Ok(best_play(observation, plays, rng))
}
